From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Herman van Hazendonk <github.com@herrie.org>
Date: Thu, 22 Nov 2018 16:40:49 +0100
Subject: [PATCH] scripts: use pkg-config to locate libcrypto

Otherwise build fails if the headers are not in the default location. While at
it also ask pkg-config for the libs, with fallback to the existing value.


MJ: modified to work in OE cross-compilation where host has
openssl-1.1 and native sysroot openssl-3.0:
1) use pkg-config-native instead of pkg-config
2) add CRYPTO_RPATH otherwise libcrypto.so.3 isn't found
   when extract-cert is used during build:

  CALL    /OE/build/luneos-kirkstone/webos-ports/tmp-glibc/work-shared/tissot/kernel-source/scripts/checksyscalls.sh
  CHK     include/generated/compile.h
  EXTRACT_CERTS   /OE/build/luneos-kirkstone/webos-ports/tmp-glibc/work-shared/tissot/kernel-source/certs/"verity.x509.pem"
  EXTRACT_CERTS   certs/signing_key.pem
scripts/extract-cert: error while loading shared libraries: libcrypto.so.3: cannot open shared object file: No such file or directory
scripts/extract-cert: error while loading shared libraries: libcrypto.so.3: cannot open shared object file: No such file or directory
rm: cannot remove 'certs/x509_certificate_list'rm: cannot remove 'certs/signing_key.x509': No such file or directory
: No such file or directory
make[3]: *** [/OE/build/luneos-kirkstone/webos-ports/tmp-glibc/work-shared/tissot/kernel-source/certs/Makefile:102: certs/signing_key.x509] Error 1

* Fixes:
martin@jama:/OE/build/luneos-kirkstone/webos-ports/tmp-glibc/work/tissot-webos-linux/linux-xiaomi-tissot/4.9.188+gitrAUTOINC+1ef326cef2-r0/build$ scripts/extract-cert certs/signing_key.pem  foo
At main.c:154:
- SSL error:0909006C:PEM routines:get_name:no start line: ../crypto/pem/pem_lib.c:745
extract-cert: certs/signing_key.pem: Success
martin@jama:/OE/build/luneos-kirkstone/webos-ports/tmp-glibc/work/tissot-webos-linux/linux-xiaomi-tissot/4.9.188+gitrAUTOINC+1ef326cef2-r0/build$ echo $?
1

it uses libcrypto from host:
martin@jama:/OE/build/luneos-kirkstone/webos-ports/tmp-glibc/work/tissot-webos-linux/linux-xiaomi-tissot/4.9.188+gitrAUTOINC+1ef326cef2-r0/build$ ldd scripts/extract-cert
        linux-vdso.so.1 (0x00007ffd19f14000)
        libcrypto.so.1.1 => /lib/x86_64-linux-gnu/libcrypto.so.1.1 (0x00007f929aa27000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f929a7ff000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f929ad30000)

extract-cert from honister works OK:
martin@jama:/OE/build/luneos-kirkstone/webos-ports/tmp-glibc/work/tissot-webos-linux/linux-xiaomi-tissot/4.9.188+gitrAUTOINC+1ef326cef2-r0/build$ /OE/build/luneos-honister/webos-ports/tmp-glibc/work/tissot-webos-linux/linux-xiaomi-tissot/4.9.188+gitrAUTOINC+1ef326cef2-r0/build/scripts/extract-cert certs/signing_key.pem  foo
martin@jama:/OE/build/luneos-kirkstone/webos-ports/tmp-glibc/work/tissot-webos-linux/linux-xiaomi-tissot/4.9.188+gitrAUTOINC+1ef326cef2-r0/build$ echo $?
0

and uses libcrypto from native sysroot:
martin@jama:/OE/build/luneos-kirkstone/webos-ports/tmp-glibc/work/tissot-webos-linux/linux-xiaomi-tissot/4.9.188+gitrAUTOINC+1ef326cef2-r0/build$ ldd /OE/build/luneos-honister/webos-ports/tmp-glibc/work/tissot-webos-linux/linux-xiaomi-tissot/4.9.188+gitrAUTOINC+1ef326cef2-r0/build/scripts/extract-cert
        linux-vdso.so.1 (0x00007ffdf17e2000)
        libcrypto.so.1.1 => /OE/build/luneos-honister/webos-ports/tmp-glibc/work/tissot-webos-linux/linux-xiaomi-tissot/4.9.188+gitrAUTOINC+1ef326cef2-r0/recipe-sysroot-native/usr/lib/libcrypto.so.1.1 (0x00007f7d02bff000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f7d029b0000)
        /OE/build/luneos-honister/webos-ports/tmp-glibc/sysroots-uninative/x86_64-linux/lib/ld-linux-x86-64.so.2 => /lib64/ld-linux-x86-64.so.2 (0x00007f7d02eb7000)

Originally by Rolf Eike Beer <eb@emlix.com>, taken from shr-distribution/linux tissot/4.9/halium-9.0 (0757ebf08557), where tissot builds with it.

Signed-off-by: Herman van Hazendonk <github.com@herrie.org>
Upstream-Status: Backport
---
 scripts/Makefile | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/scripts/Makefile b/scripts/Makefile
index 1d80897a964..9cc5f4a9137 100644
--- a/scripts/Makefile
+++ b/scripts/Makefile
@@ -11,6 +11,10 @@
 
 HOST_EXTRACFLAGS += -I$(srctree)/tools/include
 
+CRYPTO_LIBS = $(shell pkg-config-native --libs libcrypto 2> /dev/null || echo -lcrypto)
+CRYPTO_CFLAGS = $(shell pkg-config-native --cflags libcrypto 2> /dev/null)
+CRYPTO_RPATH = -Wl,-rpath,$(shell pkg-config-native --variable=libdir libcrypto 2> /dev/null)
+
 hostprogs-$(CONFIG_KALLSYMS)     += kallsyms
 hostprogs-$(CONFIG_LOGO)         += pnmtologo
 hostprogs-$(CONFIG_VT)           += conmakehash
@@ -23,8 +27,9 @@ hostprogs-$(CONFIG_SYSTEM_EXTRA_CERTIFICATE) += insert-sys-cert
 
 HOSTCFLAGS_sortextable.o = -I$(srctree)/tools/include
 HOSTCFLAGS_asn1_compiler.o = -I$(srctree)/include
-HOSTLOADLIBES_sign-file = -lcrypto
-HOSTLOADLIBES_extract-cert = -lcrypto
+HOSTLOADLIBES_sign-file = $(CRYPTO_LIBS)
+HOSTCFLAGS_extract-cert.o = $(CRYPTO_CFLAGS)
+HOSTLOADLIBES_extract-cert = $(CRYPTO_LIBS) $(CRYPTO_RPATH)
 
 always		:= $(hostprogs-y) $(hostprogs-m)
 
